### PR TITLE
fix saturating float-to-int cast

### PR DESCRIPTION
**Pull request by Rexicon226**
The issue is that Kani computes int_max_as_float by casting u64::MAX to f64. But u64::MAX (2^64 - 1) isn't exactly representable in f64, so it rounds *up* to 2^64. The old code used a greater-than operation, which would mean a float value equal to exactly 2^64 fails this check, falls through the raw truncation path, and produces a wrong result instead of saturating to u64::MAX.

Instead we use greater-than-or-equal so that the boundary value 2^64 correctly saturates. This is safe for types where the max IS exactly representable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs [hunk 1]
```diff
@@ -1212,7 +1212,7 @@ impl GotocCtx<'_, '_> {
         // 3. src < int_min -> result is MIN (or 0 for unsigned)
         // 4. Otherwise -> truncate toward zero
         let is_nan = src_expr.clone().is_nan();
-        let above_max = src_expr.clone().gt(int_max_as_float.clone());
+        let above_max = src_expr.clone().ge(int_max_as_float.clone());
```
Comment (feliperodri) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:1215]: Thank you for this fix! There is also another potential gap, `u128` and `i128` have the same problem.

The `int_bounds_as_float` function converts `u128::MAX` and `i128::MAX` to `f64` the same way:

```rust name=rvalue.rs url=https://github.com/model-checking/kani/blob/e43314b17ee625db430f7102f989e4b2ab02f6e5/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs#L1278-L1279
UintTy::U128 => u128::MAX as f64,
```

`u128::MAX` similarly rounds up to a float that is not `u128::MAX`, causing the same kind of off-by-one at the boundary. The fix in this PR (`ge` for max) should cover this case too — but it's worth adding a test for `f64 as u128` at the boundary value to confirm.

Could you also add tests for `u128` and `i64`/`i128` boundaries?

Consider adding:
```rust
#[kani::proof]
fn check_f64_to_u128_saturation() {
    // u128::MAX rounds up in f64, similar to u64::MAX
    let x: f64 = u128::MAX as f64; // = 2^128 in f64
    let y: u128 = x as u128;
    kani::assert(y == u128::MAX, "matches");
}
```

One last thing. The comment at line 1200–1202 is partially misleading

```rust name=rvalue.rs url=https://github.com/model-checking/kani/blob/e43314b17ee625db430f7102f989e4b2ab02f6e5/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs#L1198-L1204
// Note: We convert the integer bounds to float for comparison.
// For very large integer types (i128, u128), the float conversion may lose precision,
// but this is acceptable because:
// 1. Any float value that's truly larger than MAX_INT will still compare as > MAX_INT
// 2. Any float value that's truly smaller than MIN_INT will still compare as < MIN_INT
```

Point 1 is now slightly inaccurate, since it's precisely the case where a float equals `MAX_INT_as_float` (not *strictly* greater) that this PR fixes. The comment should be updated to reflect that `>=` is used intentionally to handle this rounding.

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs (hunk 1 above)
Comment (Rexicon226) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:1215]: Good points, will address this!

Comment (Rexicon226) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:1215]: How does this look? I'm happy to refactor the comments more, just tried to be minimal for now.

File: tests/kani/Cast/float_to_int_saturating.rs [hunk 2]
```diff
@@ -1,6 +1,8 @@
 // Copyright Kani Contributors
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 
+#![feature(f128)]
```
Comment (feliperodri) [inline tests/kani/Cast/float_to_int_saturating.rs:4]: why do you need this?

**Review by feliperodri — APPROVED** [review]

File: tests/kani/Cast/float_to_int_saturating.rs (hunk 2 above)
Comment (Rexicon226) [reply tests/kani/Cast/float_to_int_saturating.rs:4]: It gave me a warning that `f128` is an unstable :thinking: 
There are other tests that also add the feature gate.

Comment (feliperodri) [reply tests/kani/Cast/float_to_int_saturating.rs:4]: Could you address all CI failures?

Comment (Rexicon226) [reply tests/kani/Cast/float_to_int_saturating.rs:4]: I believe I've fixed it :)

**Comment by Rexicon226** [comment]
The remaining failures don't seem related to this PR?

**Comment by Rexicon226** [comment]
@feliperodri sorry to bother, but any progress?
